pkg/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,411 @@
+		return nil
+	}
+
+	syncKey := syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName())
+	if c.watchesGVK(gvk) {
+		_, err = c.cfClient.AddData(ctx, instance)
+		if err != nil {
+			c.syncMetricsCache.AddObject(
+				syncKey,
+				syncutil.Tags{
+					Kind:   instance.GetKind(),
+					Status: metrics.ErrorStatus,
+				},
+			)
+
+			return err
+		}
+
+		c.syncMetricsCache.AddObject(syncKey, syncutil.Tags{
+			Kind:   instance.GetKind(),
+			Status: metrics.ActiveStatus,
+		})
+		c.syncMetricsCache.AddKind(instance.GetKind())
+	}
+
+	c.tracker.ForData(instance.GroupVersionKind()).Observe(instance)
+
+	return nil
+}
+
+func (c *CacheManager) RemoveObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	if _, err := c.cfClient.RemoveData(ctx, instance); err != nil {
+		return err
+	}
+
+	// only delete from metrics map if the data removal was successful
+	c.syncMetricsCache.DeleteObject(syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName()))
+	c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+
+	return nil
+}
+
+func (c *CacheManager) wipeData(ctx context.Context) error {
+	if _, err := c.cfClient.RemoveData(ctx, target.WipeData()); err != nil {
+		return err
+	}
+
+	// reset sync cache before sending the metric
+	c.syncMetricsCache.ResetCache()
+	c.syncMetricsCache.ReportSync()
+
+	return nil
+}
+
+func (c *CacheManager) ReportSyncMetrics() {
+	c.syncMetricsCache.ReportSync()
+}
+
+func (c *CacheManager) syncGVK(ctx context.Context, gvk schema.GroupVersionKind) error {
+	u := &unstructured.UnstructuredList{}
+	u.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind + "List",
+	})
+
+	var err error
+	func() {
+		c.mu.RLock()
+		defer c.mu.RUnlock()
+
+		// only call List if we are still watching the gvk.
+		if c.watchedSet.Contains(gvk) {
+			err = c.reader.List(ctx, u)
+		}
+	}()
+
+	if err != nil {
+		return fmt.Errorf("listing data for %+v: %w", gvk, err)
+	}
+
+	for i := range u.Items {
+		if err := c.AddObject(ctx, &u.Items[i]); err != nil {
+			return fmt.Errorf("adding data for %+v: %w", gvk, err)
+		}
+	}
+
+	return nil
+}
+
+func (c *CacheManager) manageCache(ctx context.Context) {
+	// relistStopChan is used to stop any list operations still in progress
+	relistStopChan := make(chan struct{})
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-c.backgroundManagementTicker.C:
+			func() {
+				c.mu.Lock()
+				defer c.mu.Unlock()
+
+				c.wipeCacheIfNeeded(ctx)
+
+				if !c.needToList {
+					// this means that there are no changes needed
+					// such that any gvks need to be relisted.
+					// any in flight goroutines can finish relisiting.
+					return
+				}
+
+				// otherwise, spin up new goroutines to relist gvks as there has been a wipe
+
+				// stop any goroutines that were relisting before
+				// as we may no longer be interested in those gvks
+				close(relistStopChan)
```

Comment:
we are closing relistStopChan without checking whether it has already been closed, which could result in a panic.

Reply:
the invariant here should be that this codesite is the only place that opens/ closes the channel here so we should be good IIRC.